### PR TITLE
NEW Create action

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/github-actions-ci-cd/issues/36

### CODE REVIEW

File: .github/workflows/auto-tag.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+      - '*.*.*'
+jobs:
+  auto-tag:
+    name: Auto-tag
+    runs-on: ubuntu-latest
+    steps:
+      - name: Auto-tag
+        uses: silverstripe/gha-auto-tag@main
+        with:
+          ref: ${{ github.ref }}
```
Comment (emteknetnz) [inline .github/workflows/auto-tag.yml]: Remove passing in ref + sha

File: README.md [hunk 2]
```diff
@@ -1 +1,34 @@
...
+      - '*.*.*'
+jobs:
+  auto-tag:
+    name: Auto-tag
+    runs-on: ubuntu-latest
+    steps:
+      - name: Auto-tag
+        uses: silverstripe/gha-auto-tag@main
+        with:
+          ref: ${{ github.ref }}
```
Comment (emteknetnz) [inline README.md]: Remove passing in ref + sha

File: action.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    description: SHA of the tag
+
+runs:
+  using: composite
+  steps:
+    - name: Validate inputs
+      shell: bash
+      env:
+        REF: ${{ inputs.ref }}
+        SHA: ${{ inputs.sha }}
```
Comment (emteknetnz) [inline action.yml]: not nee to validate sha, though do need to validate ${{ github.ref }}

File: action.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
+name: Auto-tag
+description: Automatically delete and re-release tags so that other github action modules can use something similar to carets e.g. v0.1 or v3
+inputs:
```
Comment (emteknetnz) [inline action.yml]: Remove inputs

File: action.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        if [ "$MAJOR" == "0" ]; then
+          NEW_TAG=("v${MAJOR}.${MINOR}")
+        fi
+        echo "Tag is $NEW_TAG"
+        echo "::set-output name=tag::$NEW_TAG"
+
+    - name: Add tag to repo
+      uses: silverstripe/gha-tag-release@main
+      with:
+        sha: ${{ github.sha }}
```
Comment (emteknetnz) [inline action.yml]: Remove sha

File: action.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    - name: Generate tag name
+      id: generate_tag_name
+      shell: bash
+      env:
+        REF: ${{ github.ref }}
+      run: |
+        # refs/tags/0.1.23 => 0.1.23
+        TAG=$(echo $REF | cut -c 11-)
+        [[ $TAG =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]
```
Comment (ChloeSartorelli) [inline action.yml]:
```suggestion
    - name: Generate tag name
      id: generate_tag_name
      shell: bash
      env:
        REF: ${{ github.ref }}
      run: |
        # refs/tags/0.1.23 => 0.1.23
        TAG=$(echo $REF | cut -c 11-)
        if ! [[ "$TAG" =~ ^([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]; then
          echo "Invalid semver tag $TAG"
          exit 1
        fi
        echo "Valid semver tag $TAG"
```

I don't think it's useful to have the validation step separate - it just results in code duplication.

File: .github/workflows/auto-tag.yml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  push:
+    tags:
+      - '*.*.*'
+jobs:
+  auto-tag:
+    name: Auto-tag
+    runs-on: ubuntu-latest
+    steps:
+      - name: Auto-tag
+        uses: silverstripe/gha-auto-tag@main
```
Comment (ChloeSartorelli) [inline .github/workflows/auto-tag.yml:12]: We should declare the expected tag here shouldn't we? i.e. `@v1`?
Or will that be done (on all repositories) as a separate PR after this first load are merged (in a new `1` branch, prior to the `1.0.0` tag being created)?

File: action.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        PATCH="${BASH_REMATCH[3]}"
+        NEW_TAG="v${MAJOR}"
+        if [ "$MAJOR" == "0" ]; then
+          NEW_TAG=("v${MAJOR}.${MINOR}")
+        fi
+        echo "Tag is $NEW_TAG"
+        echo "::set-output name=tag::$NEW_TAG"
+
+    - name: Add tag to repo
+      uses: silverstripe/gha-tag-release@main
```
Comment (ChloeSartorelli) [inline action.yml:31]: Same question as above regarding `@main` vs `@v1`

File: README.md [hunk 9]
```diff
@@ -1 +1,30 @@
...
+The action will tag v2 when 2.5.0 is released and v0.3 when 0.3.5 is released
+
+The new 'v' tag will point to the same sha  e.g. the sha of v2 will equal the sha of 2.5.0
+
+This allows modules to include workflows using the @v2 or @v0.3 syntax
+
+Using the 'v' prefix to avoid confusion with the '2' branch naming convention that Silverstripe uses
+
+## Usage
+
```
Comment (ChloeSartorelli) [inline README.md:16]:
```suggestion

This action has no configuration or inputs - just add the following workflow verbatim.

```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: .github/workflows/auto-tag.yml (hunk 7 above)
Comment (emteknetnz) [reply .github/workflows/auto-tag.yml:12]: > Or will that be done (on all repositories) as a separate PR after this first load are merged (in a new 1 branch, prior to the 1.0.0 tag being created)?

Yes - follow up PRs once everything is merged

File: action.yml (hunk 8 above)
Comment (emteknetnz) [reply action.yml:31]: Same answer as above - we'll use follow up PR's to update

File: README.md (hunk 9 above)
Comment (emteknetnz) [reply README.md:16]: Added

File: action.yml (hunk 6 above)
Comment (emteknetnz) [reply action.yml]: Have combined steps

**Review by GuySartorelli — APPROVED** [review]
Looks good
Please give me permissions necessary to merge :p

File: action.yml [hunk 10]
```diff
@@ -0,0 +1,34 @@
+name: Auto-tag
+description: Automatically delete and re-release tags so that other github action modules can use something similar to carets e.g. v0.1 or v3
+
+runs:
+  using: composite
+  steps:
+    - name: Generate tag name
```
Comment (ChloeSartorelli) [inline action.yml:8]:
```suggestion

    - name: Generate tag name
```

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Actually there is one very minor formatting problem to fix. It's the only place I've seen where there's no new line before a step.

File: action.yml (hunk 10 above)
Comment (emteknetnz) [reply action.yml:8]: Done

**Review by GuySartorelli — APPROVED** [review]
Looks good

**Pull request merged**